libraries/networking/src/ExternalResource.h

```diff
@@ -0,0 +1,139 @@
+
+#include <QObject>
+#include <QUrl>
+#include <QMap>
+
+/**
+ * Flexible management for external resources
+ *
+ * With the death of the original High Fidelity and the transition of the project to a community-managed
+ * one, it became necessary to deal with that the various assets that used to be located under the
+ * highfidelity.com domain will disappear, and there won't be a fixed place for them afterwards. Data
+ * hosted by community members may not remain forever, reorganization may be necessary, people may want
+ * to run mirrors, and some might want to run without external internet access at all.
+ *
+ * This class makes it possible to deal with this in a more flexible manner: rather than hardcoding URLs
+ * all over the codebase, now it's possible to easily change where all those things are downloaded from.
+ *
+ *
+ */
+class ExternalResource : public QObject {
+    Q_OBJECT
+
+    public:
+
+    static ExternalResource *getInstance();
+    ~ExternalResource() {};
+
+    /**
+     * Bucket from which to retrieve the resource
+     *
+     * The original High Fidelity used the Public, Content and MPAssets buckets. The intention is that the
+     * community-run versions of these will keep the original data and structure, and any new additions
+     * will be done to the Assets bucket instead. This should ease the transition and ensure a clean
+     * separation.
+     */
+    enum class Bucket {
+        /** Assets that used to be in the hifi-public S3 bucket */
+        Public,
+
+        /** Assets that used to be in the hifi-content S3 bucket */
+        Content,
+
+        /** Assets that used to be in the mpassets S3 bucket (hifi marketplace) */
+        MPAssets,
+
+        /** Vircadia assets */
+        Assets
+    };
+
+    Q_ENUM(Bucket)
+
+    /**
+     * Returns the location of a resource as a QUrl
+     *
+     * Returns the location of the resource \p relative_path in bucket \p bucket
+     *
+     * @note The resulting path will be sanitized by condensing multiple instances of '/' to one.
+     * This is done for easier usage with Amazon S3 and compatible systems.
+     *
+     * @param bucket The bucket in which the resource is found
+     * @param relative_path The path of the resource within the bucket
+     * @returns The resulting URL as a QUrl
+     */
+    Q_INVOKABLE QUrl getQUrl(Bucket bucket, const QUrl& relative_path);
+
+
+    /**
+     * Returns the location of a resource as a QUrl
+     *
+     * Returns the location of the resource \p relative_path in bucket \p bucket
+     *
+     * @note The resulting path will be sanitized by condensing multiple instances of '/' to one.
+     * This is done for easier usage with Amazon S3 and compatible systems.
+     *
+     * @param bucket The bucket in which the resource is found
+     * @param relative_path The path of the resource within the bucket
+     * @returns The resulting URL as a QUrl
+     */
+    Q_INVOKABLE QUrl getQUrl(Bucket bucket, QString path) {
```

Comment:
do these all need to be Q_INVOKABLE? I think the public interface should just be getUrl, and we shouldn't expose methods that reference Qt types like QUrl

Reply:
I wasn't entirely sure there. Removed. It looks like QUrl just behaves like a string in the console anyway.